mousestyles/classification/clustering.py

```diff
@@ -0,0 +1,303 @@
+    mouse_dayavgstd_X_scl = mouse_dayavgstd_X / np.max(
+        mouse_dayavgstd_X, axis=0)
+    mouse_dayavgstd_scl = np.hstack(
+        [mouse_dayavgstd[:, 0:2], mouse_dayavgstd_X_scl])
+    if (std is False and rescale is False):
+        return(mouse_dayavg)
+    elif (std is True and rescale is True):
+        return(mouse_dayavgstd)
+    elif (std is False and rescale is True):
+        return(mouse_dayavgstd_scl[:, 0:(mouse_dayavg.shape[1])])
+    else:
+        return(mouse_dayavgstd_scl)
+
+
+# model fitting functions
+def get_optimal_hc_params(mouse_day):
+    """
+    Returns a list of 2: [method, dist]
+        method: {'ward', 'average', 'complete'}
+        dist: {'cityblock', 'euclidean', 'chebychev'}
+
+    Parameters
+    ----------
+    mouse_day: a 170 * M numpy array,
+        column 0 : strain,
+        column 1: mouse,
+        other columns corresponding to feature avg/std of a mouse over 16 days
+
+    Returns
+    -------
+    method_distance: list
+        [method, dist]
+    """
+    methods = ['ward', 'average', 'complete']
+    dists = ['cityblock', 'euclidean', 'chebychev']
+
+    method_dists = [(methods[i], dists[j]) for i in range(len(methods))
+                    for j in range(len(dists))]
+    method_dists = [(method, dist) for method, dist in method_dists
+                    if method != 'ward' or dist == 'euclidean']
+
+    cs = []
+    for method, dist in method_dists:
+        Z = linkage(mouse_day[:, 2:], method=method, metric=dist)
+        c, coph_dists = cophenet(Z, pdist(mouse_day[:, 2:]))
+        cs.append(c)
+
+    # determine the distance method
+    method, dist = method_dists[np.argmax(cs)]
+    return([method, dist])
+
+
+def fit_hc(mouse_day_X, method, dist, num_clusters=range(2, 17)):
+    """
+    Returns a list of 2: [silhouettes, cluster_labels]
+        silhouettes: list of float,
+        cluster_labels: list of list,
+            each sublist is the labels corresponding to the silhouette
+
+    Parameters
+    ----------
+    mouse_day_X: a 170 * M numpy array,
+        all columns corresponding to feature avg/std of a mouse over 16 days
+    method: str,
+        method of calculating distance between clusters
+    dist: str,
+        distance metric
+    num_clusters: range
+        range of number of clusters
+
+    Returns
+    -------
+    A list of 2: [silhouettes, cluster_labels]
+    """
+    if (dist == "chebychev"):
+        dist = "chebyshev"
+    cluster_labels = []
+    silhouettes = []
+    for n_clusters in num_clusters:
+        clustering = AgglomerativeClustering(
+            linkage=method, n_clusters=n_clusters)
+        clustering.fit(mouse_day_X)
+        labels = clustering.labels_
+        silhouettes.append(metrics.silhouette_score(
+            mouse_day_X, labels, metric=dist))
+        cluster_labels.append(list(labels))
+
+    return([silhouettes, cluster_labels])
+
+
+# plotting functions
+def plot_dendrogram(mouse_day, method, dist):
```

Comment:
Since `plot_dendrogram` and `plot_lastp_dist` are added into the visualization folder, please delete them from this script. @pttzty @MengfeiJiang